src/audio/SDL_audiotypecvt.c

```diff
@@ -39,50 +39,155 @@
 #define NEED_SCALAR_CONVERTER_FALLBACKS 1
 #endif
 
-#define DIVBY128     0.0078125f
-#define DIVBY32768   0.000030517578125f
-#define DIVBY8388607 0.00000011920930376163766f
-
 #if NEED_SCALAR_CONVERTER_FALLBACKS
 
-/* these all convert backwards because (currently) float32 is >= to the size of anything it converts to, so it lets us safely convert in-place. */
-#define AUDIOCVT_TOFLOAT_SCALAR(from, fromtype, equation) \
-    static void SDL_Convert_##from##_to_F32_Scalar(float *dst, const fromtype *src, int num_samples) { \
-        int i; \
-        LOG_DEBUG_AUDIO_CONVERT(#from, "F32"); \
-        for (i = num_samples - 1; i >= 0; --i) { \
-            dst[i] = equation; \
-        } \
-    }
-
-AUDIOCVT_TOFLOAT_SCALAR(S8, Sint8, ((float)src[i]) * DIVBY128)
-AUDIOCVT_TOFLOAT_SCALAR(U8, Uint8, (((float)src[i]) * DIVBY128) - 1.0f)
-AUDIOCVT_TOFLOAT_SCALAR(S16, Sint16, ((float)src[i]) * DIVBY32768)
-AUDIOCVT_TOFLOAT_SCALAR(S32, Sint32, ((float)(src[i] >> 8)) * DIVBY8388607)
-#undef AUDIOCVT_FROMFLOAT_SCALAR
-
-/* these all convert forwards because (currently) float32 is >= to the size of anything it converts from, so it lets us safely convert in-place. */
-#define AUDIOCVT_FROMFLOAT_SCALAR(to, totype, clampmin, clampmax, equation) \
-    static void SDL_Convert_F32_to_##to##_Scalar(totype *dst, const float *src, int num_samples) { \
-        int i; \
-        LOG_DEBUG_AUDIO_CONVERT("F32", #to); \
-        for (i = 0; i < num_samples; i++) { \
-            const float sample = src[i]; \
-            if (sample >= 1.0f) { \
-                dst[i] = (totype) (clampmax); \
-            } else if (sample <= -1.0f) { \
-                dst[i] = (totype) (clampmin); \
-            } else { \
-                dst[i] = (totype) (equation); \
-            } \
-        } \
-    }
-
-AUDIOCVT_FROMFLOAT_SCALAR(S8, Sint8, -128, 127, sample * 127.0f);
-AUDIOCVT_FROMFLOAT_SCALAR(U8, Uint8, 0, 255, (sample + 1.0f) * 127.0f);
-AUDIOCVT_FROMFLOAT_SCALAR(S16, Sint16, -32768, 32767, sample * 32767.0f);
-AUDIOCVT_FROMFLOAT_SCALAR(S32, Sint32, -2147483648LL, 2147483647, ((Sint32)(sample * 8388607.0f)) << 8);
-#undef AUDIOCVT_FROMFLOAT_SCALAR
+/* This code requires that floats are in the IEEE-754 binary32 format */
+SDL_COMPILE_TIME_ASSERT(float_bits, sizeof(float) == sizeof(Uint32));
+
+union float_bits {
+    Uint32 u32;
+    float f32;
+};
+
+static void SDL_Convert_S8_to_F32_Scalar(float *dst, const Sint8 *src, int num_samples)
+{
+    int i;
+
+    LOG_DEBUG_AUDIO_CONVERT("S8", "F32");
+
+    for (i = num_samples - 1; i >= 0; --i) {
+        /* 1) Construct a float in the range [65536.0, 65538.0)
+         * 2) Shift the float range to [-1.0, 1.0) */
+        union float_bits x = { .u32 = (Uint8)src[i] ^ 0x47800080u };
```

Comment:
I think we're avoiding this initializer syntax, since we support older C compilers still. Do this instead:

```c
union float_bits x;
x.u32 = (Uint8)src[i] ^ 0x47800080u;
```

Reply:
(But also, was this supposed to be a Uint32 cast here instead of Uint8?)

Reply:
I don't think we went full C99, just started allowing variable declarations mixed into code and `//` comments.

I'm not sure the cast does anything, since `src[i]` is already a Uint8, and the cast would apply before the `^` operation, so it's likely implicitly casting from Uint8 to Uint32 anyhow. But I could be totally wrong here.

Reply:
Also, I totally missed the hex floats...I'm willing to let those stay until a compiler complains, since it's a simple mechanical change to make them compile, if it comes to that.

Reply:
Cool, I've removed the designated initializers.

And yeah the cast was unnecessary U8, though for S8/S16 it is needed for avoiding sign-extension. They are implicitly cast to Uint32 after, but I don't think there's any point adding explicit casts for that.